### DISCUSSION TITLE
Debugging graalvm.python embedded python scripts when invoked by Java

### DISCUSSION THREAD

**Discussion by s3ts** [oracle/graalpython#422] [General]
How can I debug graalpy when invoked by Java?

I'm seeing that 24.0.2 release of graalpy doesn't support --inspect option as written in this [reference page](https://www.graalvm.org/latest/reference-manual/python/Tooling/).

If there's a way to do this, great -- but now lets say I want to add Java context to the scripting engine.

Is there not a way to add built-ins before I start this script, and then debug it?

This type of tooling is pretty solidly available for CPython and its C/C++ APIs. While I do see the advantage in working with graalpy via the archetypes in maven/gradle, I don't see much other advantages. In fact, I'm inclined to say writing big infrastructure this way and depending on this experimental stuff without proper tooling is hard to envision a future for. 

The alternatives I've already found includes:
* [JEP](https://github.com/ninia/jep)
* [Jpype](https://jpype.readthedocs.io/en/latest/)
* maybe just writing REST APIs and opening it up to the many different languages that OpenAPI Swagger, protobuf, etc. already supports.

I can see that this can go far and I am looking for Java-related technologies just like this. But I just think the advantage of implementing Python using graalvm can't be just speed. It has to be about re-using customized Java code in Python in rapid prototypes.

If it can't be done that way, then, unfortunately, being able to invoke some custom python only through mvn or gradle is not enough for me to go down this road. I need the tooling because if I write python code that is meant to run in an embedded manner, it will definitely require maintenance.

**Comment by msimacek** [comment]
GraalPy does support inspect. When you're embedding into Java, you need to add a dependency on it:
```xml
        <dependency>
            <groupId>org.graalvm.tools</groupId>
            <artifactId>chromeinspector-tool</artifactId>
            <version>${graalvm.version}</version>
        </dependency>
```
Then you can enable inspect on the context builder with a port number as an argument:
```java
.option("inspect", "5555")
```

[Here's the doc for that](https://www.graalvm.org/latest/tools/chrome-debugger/#programmatic-launch-of-inspector-backend)

All the other GraalVM tools like cpu sampler or insights should work as well, you just need to add the appropriate dependency.

**Reply by fniephaus** [reply]
You can also use the Debug Adapter Protocol (DAP) and use, for example, VS Code for debugging: https://www.graalvm.org/latest/tools/dap/
